dhall/src/Dhall/Parser/Expression.hs

```diff
@@ -474,14 +474,12 @@ parsers embedded = Parsers {..}
             alternative04 = (do
                 _openBrace
 
-                whitespace
-
-                _ <- optional (_comma *> whitespace)
+                -- If it fails to parse the _comma we need to recover the
```

Comment:
Instead of wrapping these all in `try`, I think you can parse the whitespace once (wrapped in `src`), and pass the parsed `Src` span as a function argument to each of the alternative record parses.  I'm fairly confident that will be more efficient